File: random-stuff/vuediff.rb
def update_children(old_ch, new_ch)
  old_start_idx = 0
  new_start_idx = 0
  old_end_idx = old_ch.length.pred
  new_end_idx = new_ch.length.pred

  while old_start_idx <= old_end_idx && new_start_idx <= new_end_idx
    old_start_vnode = old_ch[old_start_idx]
    old_start_idx += 1 unless old_start_vnode
    old_end_vnode = old_ch[old_end_idx]
    old_end_idx += 1 unless old_end_vnode
    new_start_vnode = new_ch[new_start_idx]
    new_end_vnode = new_ch[new_end_idx]

    if same?(old_start_vnode, new_start_vnode)
      patch_vnode(old_start_vnode, new_start_vnode)
      old_start_idx += 1
      new_start_idx += 1
      next
    end

    if same?(old_end_vnode, new_end_vnode)
      patch_vnode(old_end_vnode, new_end_vnode)
      old_end_idx -= 1
      new_end_idx -= 1
      next
    end

    if same?(old_start_vnode, new_end_vnode)
      patch_vnode(old_start_vnode, new_end_vnode)
      move(old_start_vnode, after: old_end_vnode)
      old_start_idx += 1
      new_end_idx -= 1
      next
    end

    if same?(old_end_vnode, new_start_vnode)
      patch_vnode(old_end_vnode, new_start_vnode)
      move(old_end_vnode, before: old_start_vnode)
      old_end_idx -= 1
      new_start_idx += 1
      next
    end

    old_key_to_idx = create_key_to_old_idx(old_ch, old_start_idx, old_end_idx)

    idx_in_old = new_start_vnode[:key] && old_key_to_idx[new_start_vnode[:key]]

    unless idx_in_old
      insert(init_vnode(new_start_vnode), before: old_start_vnode)
      new_start_idx += 1
      next
    end

    vnode_to_move = old_ch[idx_in_old]

    if same?(vnode_to_move, new_start_vnode)
      old_ch[idx_in_old] = nil
      move(patch_vnode(vnode_to_move, new_start_vnode), before: old_start_vnode)
      new_start_idx += 1
      next
    end

    puts "Same key but different element, treat as new element"
    insert(init_vnode(new_start_vnode), before: old_start_vnode)

    new_start_idx += 1
  end

  if old_start_idx > old_end_idx
    # TODO: something about ref elms from the new children
    new_ch
      .slice(new_start_idx..new_end_idx)
      .each { |vnode| insert(init_vnode(descriptor), before: nil) }
  elsif new_start_idx > new_end_idx
    old_ch
      .slice(old_start_idx..old_end_idx)
      .each { |vnode| remove_vnode(vnode) }
  end
end

def create_key_to_old_idx(children, start_idx, end_idx)
  start_idx.upto(end_idx).reduce({}) { |h, i| h.merge[children[i].key] = i }
end

def same?(a, b)
  a[:key] == b[:key]
end

def patch_vnode(a, b)
  puts "Patching #{a} with #{b}"
  a.merge(b)
end

def remove_vnode(vnode)
  puts "Removing #{vnode.inspect}"
end

def move(vnode, before: nil, after: nil)
  if before
    puts "move #{vnode.inspect} before #{before}"
  elsif after
    puts "move #{vnode.inspect} after #{after}"
  else
    puts "move #{vnode.inspect} last"
  end
end

old_ch = [
  { key: 1, name: "one" },
  { key: 2, name: "two" },
  { key: 3, name: "three" },
  { key: nil, name: "four" },
  { key: nil, name: "five" }
]

new_ch = [
  { key: 2, name: "two" },
  { key: 3, name: "three" },
  { key: nil, name: "four" },
  { key: 1, name: "one" }
]

update_children(old_ch, new_ch)

File: random-stuff/vuediff.js
function isUndef(v) {
  return v === undefined || v === null;
}

const nodeOps = {
  insertBefore(parentElm, oldEndVnode, oldStartVnode) {
    console.log("moving", oldEndVnode, "before", oldStartVnode);
  },
  insertAfter(parentElm, oldEndVnode, oldStartVnode) {
    console.log("moving", oldEndVnode, "after", oldStartVnode);
  },
};

function isDef(v) {
  return !isUndef(v);
}

function updateChildren(
  parentElm,
  oldCh,
  newCh,
  insertedVnodeQueue,
  removeOnly
) {
  let oldStartIdx = 0;
  let newStartIdx = 0;
  let oldEndIdx = oldCh.length - 1;
  let oldStartVnode = oldCh[0];
  let oldEndVnode = oldCh[oldEndIdx];
  let newEndIdx = newCh.length - 1;
  let newStartVnode = newCh[0];
  let newEndVnode = newCh[newEndIdx];
  let oldKeyToIdx, idxInOld, vnodeToMove, refElm;

  // removeOnly is a special flag used only by <transition-group>
  // to ensure removed elements stay in correct relative positions
  // during leaving transitions
  const canMove = !removeOnly;

  while (oldStartIdx <= oldEndIdx && newStartIdx <= newEndIdx) {
    if (isUndef(oldStartVnode)) {
      oldStartVnode = oldCh[++oldStartIdx]; // Vnode has been moved left
    } else if (isUndef(oldEndVnode)) {
      oldEndVnode = oldCh[--oldEndIdx];
    } else if (sameVnode(oldStartVnode, newStartVnode)) {
      patchVnode(
        oldStartVnode,
        newStartVnode,
        insertedVnodeQueue,
        newCh,
        newStartIdx
      );
      oldStartVnode = oldCh[++oldStartIdx];
      newStartVnode = newCh[++newStartIdx];
    } else if (sameVnode(oldEndVnode, newEndVnode)) {
      patchVnode(
        oldEndVnode,
        newEndVnode,
        insertedVnodeQueue,
        newCh,
        newEndIdx
      );
      oldEndVnode = oldCh[--oldEndIdx];
      newEndVnode = newCh[--newEndIdx];
    } else if (sameVnode(oldStartVnode, newEndVnode)) {
      // Vnode moved right
      patchVnode(
        oldStartVnode,
        newEndVnode,
        insertedVnodeQueue,
        newCh,
        newEndIdx
      );
      canMove && nodeOps.insertAfter(parentElm, oldStartVnode, oldEndVnode);
      oldStartVnode = oldCh[++oldStartIdx];
      newEndVnode = newCh[--newEndIdx];
    } else if (sameVnode(oldEndVnode, newStartVnode)) {
      // Vnode moved left
      patchVnode(
        oldEndVnode,
        newStartVnode,
        insertedVnodeQueue,
        newCh,
        newStartIdx
      );
      canMove &&
        nodeOps.insertBefore(parentElm, oldEndVnode.elm, oldStartVnode.elm);
      oldEndVnode = oldCh[--oldEndIdx];
      newStartVnode = newCh[++newStartIdx];
    } else {
      if (isUndef(oldKeyToIdx))
        oldKeyToIdx = createKeyToOldIdx(oldCh, oldStartIdx, oldEndIdx);
      idxInOld = isDef(newStartVnode.key)
        ? oldKeyToIdx[newStartVnode.key]
        : findIdxInOld(newStartVnode, oldCh, oldStartIdx, oldEndIdx);
      if (isUndef(idxInOld)) {
        // New element
        createElm(
          newStartVnode,
          insertedVnodeQueue,
          parentElm,
          oldStartVnode,
          false,
          newCh,
          newStartIdx
        );
      } else {
        vnodeToMove = oldCh[idxInOld];
        if (sameVnode(vnodeToMove, newStartVnode)) {
          patchVnode(
            vnodeToMove,
            newStartVnode,
            insertedVnodeQueue,
            newCh,
            newStartIdx
          );
          console.log(`Clearing index`, idxInOld);
          oldCh[idxInOld] = undefined;
          canMove &&
            nodeOps.insertBefore(parentElm, vnodeToMove.elm, oldStartVnode.elm);
        } else {
          // same key but different element. treat as new element
          createElm(
            newStartVnode,
            insertedVnodeQueue,
            parentElm,
            oldStartVnode,
            false,
            newCh,
            newStartIdx
          );
        }
      }
      newStartVnode = newCh[++newStartIdx];
    }
  }
  if (oldStartIdx > oldEndIdx) {
    refElm = isUndef(newCh[newEndIdx + 1]) ? null : newCh[newEndIdx + 1].elm;
    addVnodes(
      parentElm,
      refElm,
      newCh,
      newStartIdx,
      newEndIdx,
      insertedVnodeQueue
    );
  } else if (newStartIdx > newEndIdx) {
    removeVnodes(oldCh, oldStartIdx, oldEndIdx);
  }
}

function patchVnode(
  vnode,
  newStartVnode,
  insertedVnodeQueue,
  newCh,
  newStartIdx
) {
  console.log("patching vnode", vnode, "with", newStartVnode);
}

function checkDuplicateKeys(children) {
  const seenKeys = {};
  for (let i = 0; i < children.length; i++) {
    const vnode = children[i];
    const key = vnode.key;
    if (isDef(key)) {
      if (seenKeys[key]) {
        warn(
          `Duplicate keys detected: '${key}'. This may cause an update error.`,
          vnode.context
        );
      } else {
        seenKeys[key] = true;
      }
    }
  }
}

function createKeyToOldIdx(children, beginIdx, endIdx) {
  let i, key;
  const map = {};
  for (i = beginIdx; i <= endIdx; ++i) {
    key = children[i].key;
    if (isDef(key)) map[key] = i;
  }
  return map;
}

function findIdxInOld(node, oldCh, start, end) {
  for (let i = start; i < end; i++) {
    const c = oldCh[i];
    if (isDef(c) && sameVnode(node, c)) return i;
  }
}

function sameVnode(a, b) {
  return a.key === b.key && a.type === b.type;
}

function addVnodes(
  parentElm,
  refElm,
  newCh,
  newStartIdx,
  newEndIdx,
  insertedVnodeQueue
) {
  console.log(
    "adding vnodes",
    newStartIdx,
    newEndIdx,
    newCh.slice(newStartIdx, newEndIdx + 1)
  );
}

function removeVnodes(oldCh, oldStartIdx, oldEndIdx) {
  console.log(
    "removing vnodes",
    oldStartIdx,
    oldEndIdx,
    oldCh.slice(oldStartIdx, oldEndIdx + 1)
  );
}

function createElm(
  newStartVnode,
  insertedVnodeQueue,
  parentElm,
  refElm,
  idk,
  newCh,
  newStartIdx
) {
  console.log("Creating ", newStartVnode, "before", refElm);
}

const oldCh = [
  { key: 1, name: "one" },
  { key: 2, name: "two" },
  { key: 3, name: "three" },
  { key: null, name: "four" },
  { key: null, name: "five" },
];
const newCh = [
  { key: 2, name: "two" },
  { key: 3, name: "three" },
  { key: null, name: "four" },
  { key: 1, name: "one" },
];
const insertedNodeQueue = [];

updateChildren(null, oldCh, newCh, insertedNodeQueue);
